src/test/java/picard/sam/markduplicates/AbstractMarkDuplicatesCommandLineProgramTester.java

```diff
@@ -143,33 +144,33 @@ public void setExpectedOpticalDuplicate(final int expectedOpticalDuplicatePairs)
     }
 
     @Override
-    public void test() {
+    public void test() throws IOException {
         try {
             updateExpectedDuplicationMetrics();
 
             // Read the output and check the duplicate flag
             int outputRecords = 0;
             final Set<String> sequencingDTErrorsSeen = new HashSet<>();
-            final SamReader reader = SamReaderFactory.makeDefault().open(getOutput());
-            for (final SAMRecord record : reader) {
-                outputRecords++;
-                final String key = samRecordToDuplicatesFlagsKey(record);
-                if (!this.duplicateFlags.containsKey(key)) {
-                    System.err.println("DOES NOT CONTAIN KEY: " + key);
-                }
-                Assert.assertTrue(this.duplicateFlags.containsKey(key));
-                final boolean value = this.duplicateFlags.get(key);
-                this.duplicateFlags.remove(key);
-                if (value != record.getDuplicateReadFlag()) {
-                    System.err.println("Mismatching read:");
-                    System.err.print(record.getSAMString());
-                }
-                Assert.assertEquals(record.getDuplicateReadFlag(), value);
-                if (testOpticalDuplicateDTTag && MarkDuplicates.DUPLICATE_TYPE_SEQUENCING.equals(record.getAttribute("DT"))) {
-                    sequencingDTErrorsSeen.add(record.getReadName());
+            try(final SamReader reader = SamReaderFactory.makeDefault().open(getOutput())) {
+                for (final SAMRecord record : reader) {
+                    outputRecords++;
+                    final String key = samRecordToDuplicatesFlagsKey(record);
+                    if (!this.duplicateFlags.containsKey(key)) {
+                        System.err.println("DOES NOT CONTAIN KEY: " + key);
```

Comment:
should the test fail here? what does this mean? no-one looks at stderr of tests....

Reply:
This is followed by `Assert.assertTrue(this.duplicateFlags.containsKey(key));` so the check and `println` are redundant. The assert is redundant as well since the line after it calls `get()` which will throw if the key is missing anyway. I would remove lines 158-161.